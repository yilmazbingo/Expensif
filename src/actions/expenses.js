import uuid from "uuid";
import database from "../firebase/firebase";

// ADD_EXPENSE
//it has an object to be dispatched. this will actually will change the redux store
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

//this will just going to start that process off. it is going to actually dispatch addExpense
//this will return a function. this is the function that is going to  actually work since we set up the middleware for redux-thunk.
//this function internally called by redux
//first we are gonna write the data to firebase and then we will call addExpense
export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    //we returned this so we can use it in promise chaining. so we can use in async test cases
    return database
      .ref("expenses")
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});
