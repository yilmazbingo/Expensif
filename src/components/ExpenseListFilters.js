import React from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters";
import { DateRangePicker } from "react-dates";

class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null //null or string. not false or true
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(setEndDate));
  };
  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={e => {
            this.props.dispatch(setTextFilter(e.target.value));
          }}
        />
        <select
          name=""
          id=""
          value={this.props.filters.sortBy}
          onChange={e => {
            if (e.target.value === "date") {
              this.props.dispatch(sortByDate());
            } else if (e.target.value === "amount") {
              this.props.dispatch(sortByAmount());
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDateId="MyDatePickerStart"
          endDateId="MyDatePickerEnd"
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

// const ExpenseListFilters = props => (

// );

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);
//onChange takes a function and every time the input changes the function fires.
