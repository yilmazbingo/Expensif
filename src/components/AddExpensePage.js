
import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

//this is for testing
//instead of importing an action generator, we pass it via the props
//React recommend using the object form of mapDispatchToProps unless you specifically need to customize dispatching behavior in some way.

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
