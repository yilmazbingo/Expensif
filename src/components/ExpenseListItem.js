import React from "react";
import { connect } from "react-redux";
import { removeExpense } from "../actions/expenses";

const ExpenseListItem = ({ dispatch, id, description, amount, createdAt }) => (
  <div>
    <h3>{description}</h3>
    <p>
      {amount}-{createdAt}
    </p>
    <button
      onClick={() => {
        dispatch(removeExpense({ id }));
      }}
    >
      Remove
    </button>
  </div>
);

// const mapStateToProps = state => {
//   return {
//     expenses: state.expenses
//   };
// };

export default connect()(ExpenseListItem);
