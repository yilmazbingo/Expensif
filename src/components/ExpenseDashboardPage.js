import React from "react";
import ExpenseList from "./ExpeseList";
import ExpenseListFilters from "./ExpenseListFilters";
const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
