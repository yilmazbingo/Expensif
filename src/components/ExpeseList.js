import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";
const ExpenseList = props => (
  <div>
    <h1>Expense List</h1>
    {props.expenses.map(expense => {
      return <ExpenseListItem key={expense.id} {...expense} />;
    })}
  </div>
);

const mapStatetoProp = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
    filters: state.filters
  };
};
export default connect(mapStatetoProp)(ExpenseList);
//this is whre we provide the information about what we want to connect.
//component needs only a subset of the store
