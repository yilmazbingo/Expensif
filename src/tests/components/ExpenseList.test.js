import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpeseList";
import expenses from "../fixtures/expenses";
// import toJSON from "enzyme-to-json";

test("should render ExpenseList with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseList with 'no expense' title", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
