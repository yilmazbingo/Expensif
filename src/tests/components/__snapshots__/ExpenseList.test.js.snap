// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with 'no expense' title 1`] = `
<div>
  <p>
    no expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={12}
    createdAt={0}
    description="gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={100}
    createdAt={-345600000}
    description="rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={1000}
    createdAt={345600000}
    description="card"
    id="3"
    key="3"
    note=""
  />
</div>
`;
